File: Nim/NimQml/private/QGuiApplication.nim
proc dos_qguiapplication_create() {.cdecl, dynlib: dOtherSideDll, importc.}
proc dos_qguiapplication_exec() {.cdecl, dynlib:dOtherSideDll, importc.}
proc dos_qguiapplication_quit() {.cdecl, dynlib:dOtherSideDll, importc.}
proc dos_qguiapplication_delete() {.cdecl, dynlib:dOtherSideDll, importc.}

proc create*(application: QGuiApplication) = 
  ## Create a new QApplication
  dos_qguiapplication_create()
  application.deleted = false

proc exec*(application: QGuiApplication) =
  ## Start the Qt event loop
  dos_qguiapplication_exec()

proc quit*(application: QGuiApplication) =
  ## Quit the Qt event loop  
  dos_qguiapplication_quit()
  
proc delete*(application: QGuiApplication) = 
  ## Delete the given QApplication
  if not application.deleted:
    debugMsg("QApplication", "delete")
    dos_qguiapplication_delete()
    application.deleted = true

proc newQGuiApplication*(): QGuiApplication =
  ## Return a new QApplication  
  newWithCondFinalizer(result, delete)
  result.create()

File: lib/Gui/src/QGuiApplication.cpp
#include "DOtherSide/Gui/Gui.h"

#include <QGuiApplication>

void dos_qguiapplication_create()
{
    static int argc = 1;
    static char empty[1] = {0};
    static char *argv[] = {empty};

    register_meta_types();

    new QGuiApplication(argc, argv);
}

void dos_qguiapplication_delete()
{
    delete qGuiApp;
}

void dos_qguiapplication_exec()
{
    qGuiApp->exec();
}

void dos_qguiapplication_quit()
{
    qGuiApp->quit();
}
